sstables/sstables.cc

```diff
@@ -2230,6 +2230,18 @@ future<> sstable::move_to_new_dir(sstring new_dir, generation_type new_generatio
     _generation = std::move(new_generation);
 }
 
+future<> sstable::filesystem_storage::to_normal(const sstable& sst, generation_type gen, delayed_commit_changes* delay_commit) {
+    return move(sst, fs::path(dir).parent_path().native(), gen, delay_commit);
+}
+
+future<> sstable::move_from_staging(delayed_commit_changes* delay_commit) {
+    if (!requires_view_building()) {
+        return make_exception_future<>(malformed_sstable_exception("sstable is expected to be in staging"));
+    }
+
+    return _storage.to_normal(*this, generation(), delay_commit);
+}
```

Comment:
This is inversion of roles. Instead of sstable being a building block and the control plane deciding where to put it, now the sstable knows about view building and the staging directory.

Reply:
Agree. The move-to-state call should work better.